src/shred_network/duplicate_shred_listener.zig

```diff
@@ -0,0 +1,794 @@
+        self.consumed.deinit(self.allocator);
+    }
+
+    pub fn handle(self: *DuplicateShredHandler, dup_shred_data: DuplicateShredData) !void {
+        self.cacheRootInfo();
+        self.maybePruneBuffer();
+        try self.handleShredData(dup_shred_data);
+    }
+
+    fn cacheRootInfo(self: *DuplicateShredHandler) void {
+        const new_root = self.params.ledger_reader.maxRoot();
+        if (new_root == self.last_root) return;
+        self.last_root = new_root;
+        const epoch = self.params.epoch_schedule.getEpoch(self.last_root);
+        self.cached_slots_in_epoch = self.params.epoch_schedule.getSlotsInEpoch(epoch);
+    }
+
+    fn shouldConsumeSlot(self: *DuplicateShredHandler, slot: Slot) !bool {
+        const max_slot = self.last_root +| self.cached_slots_in_epoch;
+        if (!(slot > self.last_root and slot < max_slot)) return false;
+        // Returns false if a duplicate proof is already ingested for the slot,
+        // and updates local `consumed` cache with blockstore.
+        const gop = try self.consumed.getOrPut(self.allocator, slot);
+        if (!gop.found_existing) {
+            gop.value_ptr.* = try self.params.ledger_reader.isDuplicateSlot(slot);
+        }
+        return !gop.value_ptr.*;
+    }
+
+    fn maybePruneBuffer(self: *DuplicateShredHandler) void {
+        if (self.dup_buffer.count() < BUFFER_CAPACITY * 2) return;
+
+        var counts: std.AutoHashMapUnmanaged(Pubkey, usize) = .empty;
+        defer counts.deinit(self.allocator);
+        var keys_to_remove = std.ArrayListUnmanaged(Key){};
+        defer keys_to_remove.deinit(self.allocator);
+
+        var it = self.dup_buffer.keyIterator();
+        while (it.next()) |key_ptr| {
+            const key = key_ptr.*;
+            var keep =
+                key.slot > self.last_root and key.slot < self.last_root + self.cached_slots_in_epoch;
+            if (keep) {
+                keep = (self.shouldConsumeSlot(key.slot) catch false);
+            }
+            if (keep) {
+                const g = counts.getOrPut(self.allocator, key.from) catch {
+                    keys_to_remove.append(self.allocator, key) catch {};
+                    continue;
+                };
+                if (!g.found_existing) g.value_ptr.* = 0;
+                g.value_ptr.* +%= 1;
+                if (g.value_ptr.* > MAX_NUM_ENTRIES_PER_PUBKEY) {
+                    keys_to_remove.append(self.allocator, key) catch {};
+                }
+            } else {
+                keys_to_remove.append(self.allocator, key) catch {};
+            }
+        }
+        for (keys_to_remove.items) |k| {
+            _ = self.dup_buffer.remove(k);
+        }
+
+        if (self.dup_buffer.count() < BUFFER_CAPACITY) return;
+
+        var tmp = std.ArrayListUnmanaged(struct { u64, Key }){};
+        defer tmp.deinit(self.allocator);
+        var it2 = self.dup_buffer.keyIterator();
+        while (it2.next()) |key_ptr| {
+            const key = key_ptr.*;
```

Comment:
Same comment about using `keys()` instead of the iterator (and everything else).
```suggestion
        for (self.dup_buffer.keys()) |key| {
```